src/NUnitFramework/framework/Internal/Commands/TimeoutCommand.cs

```diff
@@ -0,0 +1,83 @@
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace NUnit.Framework.Internal.Commands
+{
+    using Execution;
+    using Interfaces;
+
+    /// <summary>
+    /// SetUpTearDownCommand runs SetUp methods for a suite,
+    /// runs the test and then runs TearDown methods.
+    /// </summary>
+    public class TimeoutCommand : BeforeAndAfterTestCommand
+    {
+        Timer _commandTimer;
+        private bool _commandTimedOut = false;
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SetUpTearDownCommand"/> class.
+        /// </summary>
+        /// <param name="innerCommand">The inner command</param>
+        /// <param name="timeout">Timeout value</param>
+        public TimeoutCommand(TestCommand innerCommand, int timeout)
+            : base(innerCommand)
+        {
+            Guard.ArgumentValid(innerCommand.Test is TestMethod, "SetUpTearDownCommand may only apply to a TestMethod", "innerCommand");
+            Guard.ArgumentValid(timeout > 0, "Timeout value must be greater than zero", "timeout");
+
+            BeforeTest = (context) =>
+            {
+                var testThread = Thread.CurrentThread;
+
+                // Create a timer to cancel the current thread
+                _commandTimer = new Timer(
+                    (o) =>
+                    {
+                        _commandTimedOut = true;
+                        testThread.Abort();
+                        // No join here, since the thread doesn't really terminate
+                    },
+                    null,
+                    timeout,
+                    Timeout.Infinite);
+            };
+
+            AfterTest = (context) =>
+            {
+                _commandTimer.Dispose();
+
+                // If the timer cancelled the current thread, change the result
+                if (_commandTimedOut)
+                {
+                    context.CurrentResult.SetResult(ResultState.Failure,
+                        string.Format("Test exceeded Timeout value of {0}ms", timeout));
```

Comment:
Suggestion: Could we add a test for this failure message, in case it ever gets caught up with "Cancelled by user" or something similar again?